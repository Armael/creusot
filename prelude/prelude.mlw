module CreusotPrelude
  type borrowed 'a = { current : 'a; final: 'a }

  let function ( *_ ) x = x.current
  let function ( ^_ ) x = x.final

  val borrow_mut (a : 'a) : borrowed 'a
    ensures { * result = a }
end
