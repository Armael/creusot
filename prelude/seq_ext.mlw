module SeqExt
use seq.Seq

function subsequence (s : seq 't) (i : int) (j : int) : seq 't = s[i..j]

function mk_empty () : seq 't = Seq.empty
end