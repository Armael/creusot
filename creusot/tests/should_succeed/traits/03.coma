
module C03_Impl0_F
  let%span s030 = "../03.rs" 10 8 10 9
  
  use prelude.prelude.Borrow
  
  use prelude.prelude.Intrinsic
  
  use prelude.prelude.Int32
  
  use prelude.prelude.Int
  
  let rec f (self:int32) (return'  (ret:int32))= (! bb0
    [ bb0 = s0 [ s0 =  [ &_0 <- [%#s030] (0 : int32) ] s1 | s1 = return' {_0} ]  ]
    ) [ & _0 : int32 = any_l () ]  [ return' (result:int32)-> (! return' {result}) ] 
end
module C03_Impl1_G
  let%span s030 = "../03.rs" 21 8 21 9
  
  use prelude.prelude.Borrow
  
  use prelude.prelude.Intrinsic
  
  use prelude.prelude.UInt32
  
  use prelude.prelude.Int
  
  let rec g (self:uint32) (return'  (ret:uint32))= (! bb0
    [ bb0 = s0 [ s0 =  [ &_0 <- [%#s030] (1 : uint32) ] s1 | s1 = return' {_0} ]  ]
    ) [ & _0 : uint32 = any_l () ]  [ return' (result:uint32)-> (! return' {result}) ] 
end
module C03_Impl2_H
  type g
  
  let%span s030 = "../03.rs" 30 12 30 13
  
  let%span s031 = "../03.rs" 30 22 30 24
  
  let%span span2 = "../../../../../creusot-contracts/src/invariant.rs" 27 8 27 18
  
  predicate invariant'1 (self : g)
  
  predicate inv'1 (_1 : g)
  
  axiom inv'1 [@rewrite] : forall x : g . inv'1 x = true
  
  use prelude.prelude.Borrow
  
  predicate invariant'0 (self : g) =
    [%#span2] inv'1 self
  
  predicate inv'0 (_1 : g)
  
  axiom inv'0 [@rewrite] : forall x : g . inv'0 x = true
  
  use prelude.prelude.Intrinsic
  
  let rec h (y:g) (return'  (ret:g))= {[%#s030] inv'0 y}
    (! bb0 [ bb0 = s0 [ s0 =  [ &_0 <- y ] s1 | s1 = return' {_0} ]  ] ) [ & _0 : g = any_l () | & y : g = y ] 
    [ return' (result:g)-> {[@expl:postcondition] [%#s031] inv'0 result} (! return' {result}) ]
    
end
module C03_Impl0
  let%span s030 = "../03.rs" 9 4 9 23
  
  let%span span1 = "../../../../../creusot-contracts/src/invariant.rs" 8 8 8 12
  
  let%span span2 = "../../../../../creusot-contracts/src/invariant.rs" 27 8 27 18
  
  use prelude.prelude.Int32
  
  predicate invariant'1 (self : int32) =
    [%#span1] true
  
  predicate inv'1 (_1 : int32)
  
  axiom inv'1 [@rewrite] : forall x : int32 . inv'1 x = true
  
  use prelude.prelude.Borrow
  
  predicate invariant'0 (self : int32) =
    [%#span2] inv'1 self
  
  predicate inv'0 (_1 : int32)
  
  axiom inv'0 [@rewrite] : forall x : int32 . inv'0 x = true
  
  use prelude.prelude.Int
  
  goal f_refn : [%#s030] forall self : int32 . inv'0 self  -> (forall result : int32 . inv'1 result)
end
module C03_Impl1
  let%span s030 = "../03.rs" 20 4 20 23
  
  let%span span1 = "../../../../../creusot-contracts/src/invariant.rs" 8 8 8 12
  
  let%span span2 = "../../../../../creusot-contracts/src/invariant.rs" 27 8 27 18
  
  use prelude.prelude.UInt32
  
  predicate invariant'1 (self : uint32) =
    [%#span1] true
  
  predicate inv'1 (_1 : uint32)
  
  axiom inv'1 [@rewrite] : forall x : uint32 . inv'1 x = true
  
  use prelude.prelude.Borrow
  
  predicate invariant'0 (self : uint32) =
    [%#span2] inv'1 self
  
  predicate inv'0 (_1 : uint32)
  
  axiom inv'0 [@rewrite] : forall x : uint32 . inv'0 x = true
  
  use prelude.prelude.Int
  
  goal g_refn : [%#s030] forall self : uint32 . inv'0 self
   -> (forall result : uint32 . inv'1 result /\ result = result)
end
module C03_Impl2
  type g
  
  let%span s030 = "../03.rs" 30 4 30 24
  
  let%span span1 = "../../../../../creusot-contracts/src/invariant.rs" 27 8 27 18
  
  predicate invariant'1 (self : g)
  
  predicate inv'1 (_1 : g)
  
  axiom inv'1 [@rewrite] : forall x : g . inv'1 x = true
  
  use prelude.prelude.Borrow
  
  predicate invariant'0 (self : g) =
    [%#span1] inv'1 self
  
  predicate inv'0 (_1 : g)
  
  axiom inv'0 [@rewrite] : forall x : g . inv'0 x = true
  
  goal h_refn : [%#s030] forall x : g . inv'0 x  -> inv'0 x /\ (forall result : g . inv'0 result  -> inv'0 result)
end
