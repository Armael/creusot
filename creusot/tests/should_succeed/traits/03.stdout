module Type
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
end
module CreusotContracts_Builtins_Resolve
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  use Type
  type self   
  predicate resolve (self : self)
end
module CreusotContracts_Builtins_Impl13_Resolve
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  use Type
  type t   
  predicate impl (self : t) = 
    true
end
module CreusotContracts_Builtins_Impl13
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  use Type
  type t   
  clone CreusotContracts_Builtins_Impl13_Resolve as Resolve0 with type t = t
  clone CreusotContracts_Builtins_Resolve with type self = t, predicate resolve = Resolve0.impl
end
module 03_Impl0_F
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  use Type
  clone CreusotContracts_Builtins_Impl13_Resolve as Resolve0 with type t = int32
  let rec cfg impl (o_self_1 : int32) : int32 = 
  var _0 : int32;
  var self_1 : int32;
  {
    self_1 <- o_self_1;
    goto BB0
  }
  BB0 {
    _0 <- (0 : int32);
    assume { Resolve0.impl self_1 };
    return _0
  }
  
end
module 03_Impl1_G
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  use Type
  clone CreusotContracts_Builtins_Impl13_Resolve as Resolve0 with type t = uint32
  let rec cfg impl (o_self_1 : uint32) : uint32
    ensures { result = result }
    
   = 
  var _0 : uint32;
  var self_1 : uint32;
  {
    self_1 <- o_self_1;
    goto BB0
  }
  BB0 {
    _0 <- (1 : uint32);
    assume { Resolve0.impl self_1 };
    return _0
  }
  
end
module Core_Marker_Sized
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  use Type
  type self   
end
module 03_Impl2_H
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  use Type
  type g   
  clone Core_Marker_Sized as Sized1 with type self = g
  clone CreusotContracts_Builtins_Impl13_Resolve as Resolve0 with type t = g
  let rec cfg impl (o_y_1 : g) : g = 
  var _0 : g;
  var y_1 : g;
  {
    y_1 <- o_y_1;
    goto BB0
  }
  BB0 {
    assume { Resolve0.impl _0 };
    _0 <- y_1;
    assume { Resolve0.impl y_1 };
    return _0
  }
  
end
